### ISSUE TITLE
Proposal: changes on API to handle normalizr schema

### ISSUE THREAD

**Issue by lemes** [lemes/redux-shelf#20]
Given the approaches described bellow, let's discuss over and decide how we should handle `schema`. My idea is to use [Paul Armstrong's normalizr](https://github.com/paularmstrong/normalizr) lib to give us schema support, where one could provide the defined schema to an action as an argument so that internally we use to normalize the entry payload. Later the normalized payload would be dispatched so that our reducer could handle the changes.

### Main changes

1. Forget about `ids` and `content` and go with `normalizr entities` and `result`. Relationships between entity schemas will give the users the arrays of ids. Also, we cannot predict and cover all the possible scenarios against a data list such as a server sorted and paginated list of items;

2. `set` and `update` entities actions must also accept a schema object from `normalizr` lib, e.g: `actions.update(string|schema, data)`. If a string is given, the user must provide the `data` already formated;

3. One entity action will be able to update one or more entities on store;

4. Remove all not-more-needed APIs (`normalize, idsOf ...`);

### To be considered 

1. Do not bundle `normalizr` together with the library. The user will still have the option to not use the schema;

2. We do not care whether the relationships are valid;

### Examples

**Schemas**
```javascript
import { schema } from 'redux-shelf';

// Define a users schema
export const user = new schema.Entity('users');

// Define your comments schema
export const comment = new schema.Entity('comments', {
  commenter: user
});

// Define your article 
export const article = new schema.Entity('articles', {
  author: user,
  comments: [comment]
});
```

**Action**
```javascript
import { entities, communication } from 'redux-shelf'
import { article as articleSchema } from './schemas';

export function fetchArticleById(articleId) {
  return async (dispatch) => {
    dispatch(communication.starting('article', articleId));

    try {
      const request = await fetch(`my-api/article/${articleId}`);
      const payload = await request.json();

      dispatch(entities.update(articleSchema, payload));

      dispatch(communication.done('article', articleId));
    } catch (e) {
      dispatch(communication.fail('article', articleId, e));
    }
  };
}
```

**JSON payload from api**
```
{
  "id": "123",
  "author": {
    "id": "1",
    "name": "Paul"
  },
  "title": "My awesome blog post",
  "comments": [
    {
      "id": "324",
      "commenter": {
        "id": "2",
        "name": "Nicole"
      }
    }
  ]
}
```

Dispatched payload normalirzed through your provided schema. Entities reducer will iterate over `entities` keys and update/set each entity. `result` key will be ignored and therefore not store. However, as users can still provide they own custom reducers, they are able to do whatever the heck they want with it.
```
{
  result: "123",
  entities: {
    "articles": { 
      "123": { 
        id: "123",
        author: "1",
        title: "My awesome blog post",
        comments: [ "324" ]
      }
    },
    "users": {
      "1": { "id": "1", "name": "Paul" },
      "2": { "id": "2", "name": "Nicole" }
    },
    "comments": {
      "324": { id: "324", "commenter": "2" }
    }
  }
}
```

**Comment by guilhermespopolin** [comment]
For you, what is the major benefit of using this aproach?

**Comment by lemes** [comment]
@guilhermespopolin the major benefit of this approach is that it allows users to use our API with whatever interface of data they may have. Whether they get the data from REST API, web sockets or fancy stuff like Firebase or whether the response is a nested JSON, it will work after normalizing based on a schema.

**Comment by guilhermespopolin** [comment]
I'm going to work on this. However, I have some questions:

- In your example of schema definition above you had import `schema` object from
`redux-shelf` library. If we intent do not bundle `normalizr` library with ours that shouldn't
be possible, right? 
-  I liked the approach of normalizing data response on action creator layer
(`actions.update(string|schema, data)`). So, in order to import and use `normalizr`'s
`normalize` funciton we'll declare it as a `peerDependency`?

**Comment by lemes** [comment]
Wrong, it's possible, but you have a point. `normalizr` would be a dependency of `redux-shelf` that would require to be either installed together (`peerDependencies`) or explicit added in the client owns dependencies (like you do with react and react-dom). Thus, it doesn't need to be bundle together but rather provided (existent on the node module or as global scope var, etc, depends on how we bundle). However, exporting from `redux-shelf` does kind of dismiss the possibility for clients to use their own normalisation functions (some API response might be already normalised). 

So.. I don't know. I'm open for suggestion. Do we enforce use of `normalizr`? Do we provide a way to plugin in custom normaliser?

**Comment by guilhermespopolin** [comment]
Sorry. Miss click! (facepalm)

**Comment by lemes** [comment]
lol
